src/jit/morph.cpp

```diff
@@ -9152,6 +9152,23 @@ GenTreePtr          Compiler::fgMorphSmpOp(GenTreePtr tree, MorphAddrContext* ma
 
     ASSIGN_HELPER_FOR_MOD:
 
+        // For "val % 1", return 0 if op1 doesn't have any side effects
+        // and we are not in the CSE phase, we cannot discard 'tree'
+        // because it may contain CSE expressions that we haven't yet examined.
+        //
+        if (((op1->gtFlags & GTF_SIDE_EFFECT) == 0) && !optValnumCSE_phase)
+        {
+            if (((op2->gtOper == GT_CNS_INT) && (op2->gtIntConCommon.IconValue() == 1))
+                || ((op2->gtOper == GT_CNS_LNG) && (op2->gtIntConCommon.LngValue() == 1)))
+            {
+                GenTreePtr zeroNode = new(this, GT_CNS_INT) GenTreeIntCon(typ, 0);
```

Comment:
There is no state that needs to be updated (during morph the next/prev links aren't updated - either they do not yet exist (global morph) or the caller needs to explicitly reorder them). However, it's a good idea to call DEBUG_DESTROY_NODE to trash the fields of a node that should no longer have any references.